### PR TITLE
staking: Flexible generation of reward curve and associated tweaks

### PR DESCRIPTION
**Pull request by gavofyork**
Breaks the API, sadly, but it's easy to fix your impl. Just change the line:

```
type RewardCurve = RewardCurve;
```

into:

```
type EraPayout = pallet_staking::ConvertCurve<RewardCurve>;
```

Also tweaks to `per_things.rs`:
- `from_rational_approximation` becomes `from_rational`
- `from_fraction` becomes `from_float`
- Improve efficiency of `pow`
- Implement several `ops` traits
- Ensure that `mul` is correctly characterised as non-overflowing

Also some minor refactoring to the Gilt pallet to avoid collapsing into `u128`.

Polkadot companion: paritytech/polkadot#2610

### CODE REVIEW

**Comment by gavofyork** [comment]
cc @kianenigma

File: frame/gilt/src/lib.rs [hunk 1]
```diff
@@ -96,7 +96,13 @@ pub mod pallet {
...
 
 		/// Currency type that this works on.
-		type Currency: ReservableCurrency<Self::AccountId>;
+		type Currency: ReservableCurrency<Self::AccountId, Balance=Self::CurrencyBalance>;
+
+		/// Just the `Currency::Balance` type; we have this item to allow us to constrain it to
+		/// `From<u64>`.
+		type CurrencyBalance:
+			sp_runtime::traits::AtLeast32BitUnsigned + codec::FullCodec + Copy
+				+ MaybeSerializeDeserialize + sp_std::fmt::Debug + Default + From<u64>;
```
Comment (kianenigma) [inline frame/gilt/src/lib.rs:105]: The only alternative that I can think of is to add a `where <T::Current as _>::Balance: From<u64>` to most functions of this pallet, and while this is not pretty either, I guess it has less footprint than sprinkling `where`s.

File: frame/gilt/src/lib.rs [hunk 2]
```diff
@@ -488,7 +493,40 @@ pub mod pallet {
...
+			let total_issuance = T::Currency::total_issuance();
+			let nongilt_issuance = total_issuance.saturating_sub(totals.frozen);
+			let effective_issuance = totals.proportion.left_from_one()
+				.saturating_reciprocal_mul(nongilt_issuance);
+
+			IssuanceInfo {
+				reserved: totals.frozen,
+				non_gilt: nongilt_issuance,
+				effective: effective_issuance,
+			}
```
Comment (kianenigma) [inline frame/gilt/src/lib.rs:527]:
```suggestion
			let non_gilt = total_issuance.saturating_sub(totals.frozen);
			let effective = totals.proportion.left_from_one()
				.saturating_reciprocal_mul(nongilt_issuance);

			IssuanceInfo {
				reserved: totals.frozen,
				non_gilt,
				effective,
			}
```

Looks nicer but ignore if you want

File: frame/gilt/src/mock.rs [hunk 3]
```diff
@@ -103,6 +103,7 @@ ord_parameter_types! {
 impl pallet_gilt::Config for Test {
 	type Event = Event;
 	type Currency = Balances;
+	type CurrencyBalance = u64;
```
Comment (kianenigma) [inline frame/gilt/src/mock.rs:106]:
```suggestion
	type CurrencyBalance = <Self as pallet_balances::Config>::Balance;
```

File: frame/staking/src/lib.rs [hunk 4]
```diff
@@ -782,6 +782,36 @@ impl<T: Config> SessionInterface<<T as frame_system::Config>::AccountId> for T w
 	}
 }
 
+pub trait EraPayout<Balance> {
```
Comment (kianenigma) [inline frame/staking/src/lib.rs:785]: undocumented public item.

Comment (kianenigma) [inline frame/staking/src/lib.rs:785]: Also, would be much appreciated if we provide a dummy `impl EraPayout for ()` for testing and get rid of all the verbose `pallet_staking_reward_curve::build!` in testing of pallets that won't care about this, such as babe and session. 

(Something that I've also requested previously from @thiolliere)

File: primitives/npos-elections/src/helpers.rs [hunk 5]
```diff
@@ -95,15 +95,15 @@ mod tests {
 			Assignment {
 				who: 1u32,
 				distribution: vec![
-					(10u32, Perbill::from_fraction(0.5)),
-					(20, Perbill::from_fraction(0.5)),
+					(10u32, Perbill::from_float(0.5)),
```
Comment (kianenigma) [inline primitives/npos-elections/src/helpers.rs:98]: Would be good to ensure new names are consistent with `FixedPointNumber`. I think the only one needing revision os this ^^, it is still called `from_fraction` there. Can also be a follow up.

**Review by kianenigma — APPROVED** [review]
Some naming and doc notes worth fixing here and there, but nothing major.

File: .gitignore [hunk 6]
```diff
@@ -24,3 +24,4 @@ rls*.log
 .cargo/
 .cargo-remote.toml
 *.bin
+*.iml
```
Comment (kianenigma) [inline .gitignore:27]:
```suggestion
*.iml

```

**Review by thiolliere — APPROVED** [review]

File: primitives/npos-elections/src/helpers.rs (hunk 5 above)
Comment (shawntabrizi) [reply primitives/npos-elections/src/helpers.rs:98]: @kianenigma please open an issue

**Review by shawntabrizi — APPROVED** [review]
lgtm

File: frame/staking/src/lib.rs [hunk 7]
```diff
@@ -782,6 +782,51 @@ impl<T: Config> SessionInterface<<T as frame_system::Config>::AccountId> for T w
...
+	fn era_payout(
+		_total_staked: Balance,
+		_total_issuance: Balance,
+		_era_duration_millis: u64,
+	) -> (Balance, Balance) {
+		(Default::default(), Default::default())
+	}
+}
+
+pub struct ConvertCurve<T>(sp_std::marker::PhantomData<T>);
```
Comment (shawntabrizi) [inline frame/staking/src/lib.rs:808]: probably could use a description

**Comment by gavofyork** [comment]
bot merge

**Pull request merged**